### PR TITLE
Bunch of IC Fixes

### LINKED ISSUE #456: IC inputs/outputs are ordered incorrectly
The real circuit is ordered (top-to-bottom) R E S on the left and Q, Q' on the right, while the IC has them ordered S E R and Q' Q, respectively. Not sure if it reliably reverses their order or if it is unpredictable/caused by some other criteria.

image

### LINKED ISSUE #470: IC Cullbox on Creator is incorrect

### LINKED ISSUE #639: IC's remembering previous titles
**Describe the bug**
When making multiple IC's, upon subsequent creations the previous name will still be in the title box (despite not being shown in the preview or the actual created IC unless the user updates the title in some way (deletes/adds at least one character)). If the user replaces it with a new shorter name, the created IC will still widen to the width of the longer, original name (although it will display the correct, new name).

**To Reproduce**
1. Create some valid grouping of components that can form an IC (my go-to being two switches going into an xor gate going into an LED)
2. Select the components, and choose the "Create IC" option
3. Give the IC some title, such as "Testing Title", and confirm
4. Repeat steps 1 & 2 for some new grouping of components
5. Observe the previous title is already in the "IC Name" box, and the preview has already fit the width of _that_ title.
6. Put in a new, shorter name, such as "Test Title", and confirm
7. Observe that the created IC is the same width as the first IC, despite having a shorter name

**Expected behavior**
The second time an IC is created, the title input box should still display "IC Name", and the preview should be the default width. Upon creating a new IC with a shorter name than the previous IC, it should adjust its width to the current title, not the previous one.

**Screenshots**
Example canvas:

After selecting the upper group of components and assigning them the title "Testing Title":

After selecting the lower group of components, _before_ adding anything to the title box:

What the two resulting IC's look like next to one another (after changing the second one's title to "Test Title"):

What we would have expected to appear when opening the "Create IC" menu for the lower group of components:

What we would have expected the resulting IC's width to look like:

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser: chrome

**Additional context**
Also visible in screenshot 2 above is some strange behavior when filling in the title field of the IC, in that the text expands in the preview without expanding the IC itself, resulting in some overlapping text. But I will document that further in a later issue

### LINKED ISSUE #644: IC title displaying incorrectly during creation
**Describe the bug**
When in the Create IC popup, as the user edits the "IC Name" box, the text immediately gets added to the display, without resizing it at all.

**To Reproduce**
1. Create any collection of components that form a valid IC.
2. Highlight the nodes and choose the "Create IC" option.
3. Select the "IC Name" textbox. Add many characters and observe the change in the display (I like to hold down the letter t)

**Expected behavior**
Based on the behavior on opencircuits.io, we would expect the display to not update until the user clicks out of the text box, at which point the IC box resizes at the same time as adding the title text. Alternatively, it would also make sense if the visuals updated as the user typed in the name, so long as it also resized to be large enough that text does not overlap/leave the box's edges.

**Screenshots**
What the text looks like when it gets wide enough:
title_text

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser: chrome

**Additional context**
The displayed box _will_ resize upon the user attempting to move any of the input/output nodes on the edges of the IC

### LINKED ISSUE #712: Copy/Paste duplicates ICData
**Describe the bug**
Copying then pasting an IC duplicates its ICData on the ItemNav

**To Reproduce**
1. Go to OpenCircuits
2. Load AND Gate example circuit or any other circuit that can turn into an IC
3. Select the circuit and create an IC
4. Copy the IC with CTRL+C
5. Paste the IC with CTRL+V
6. Notice that in the ItemNav, there are now 2 instances of the ICData

**Expected behavior**
Pasting ICs that have identical features shouldn't duplicate them

**Additional context**
The solution here should be entirely fixable within [`DigitalPaste`](https://github.com/OpenCircuits/OpenCircuits/blob/master/src/site/pages/digital/src/utils/DigitalPaste.ts). What needs to happen is we need to check if the incoming ICData from the paste-data is identical in structure to any of the existing ICDatas within `designer`.
This will require some function called like `ICDataEqual(ic1, ic2)` that will manually have to check if they are identical. Could potentially use serialeazy to serialize them and check if they are identical strings, but that might have problems if there's a reference to the `CircuitDesigner`.

### LINKED ISSUE #888: Pressing enter when creating IC results in error
**Describe the bug**
If the user presses enter while in the creation menu for an IC (and the title text field is _not_ selected), the IC will not be created and no more ICs will be able to be created until the page is refreshed.

**To Reproduce**
1. Place some components on the canvas
2. Select the components, and choose Create IC
3. Press enter
4. Observe the error, and attempting to click create IC (even if closing and opening again, even with new components) doing nothing

**Expected behavior**
I would expect pressing enter to function as if the user had clicked Confirm, and to not prevent ICs from being created.

**Screenshots**
The error that comes up (not a crash):
image

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser: Chrome

### LINKED ISSUE #907: ICs inside other ICs are not recognized by certain actions
**Describe the bug**
When using an IC that itself contains an IC, that contained IC isn't recognized when performing certain actions. This has been found to be an issue in two certain actions: Deleting ICData and copy/pasting between windows.

**To Reproduce (Deleting ICData)**
1. Create an IC, call it `First`.
2. Create an IC that contains `First` and call it `Second`.
3. Delete the instance of `First` on the canvas.
4. Delete `First` from the list of ICs in the itemnav.

**Expected behavior (Deleting ICData)**
When trying to delete the IC from the itemnav, it should perform the same notification to the user that the IC is in use as it would if the IC were directly placed on the campus (or a modified version to indicate what IC it is inside of).

**To Reproduce (Deleting Copy/Pasting)**
1. Create an IC, call it `First`.
2. Create an IC that contains `First` and call it `Second`.
3. Copy `Second` and paste it to a different instance of OpenCircuits.

**Expected behavior (Deleting Copy/Pasting)**
`First` and `Second` should be added to the list of ICs in the itemnav, instead only `Second` is.

### PR DESCRIPTION
**Pull request by LeonMontealegre**
Fixes #644 
Fixes #639 
Fixes #470 
Fixes #712 
Fixes #888 
Fixes #456 
Fixes #907

### CODE REVIEW

**Comment by TGCrystal** [comment]
Just tested all these issues, it seems like this fixes them!

File: src/app/digital/models/ioobjects/other/ICData.ts [hunk 1]
```diff
@@ -191,21 +191,26 @@ export class ICData {
...
+        }
+        inputs.sort(sortByPos);
+        outputs.sort(sortByPos);
 
-        const inputs = copies.getInputs().filter((i) => INPUT_WHITELIST.some((type) => i instanceof type));
-        const others = copies.getOthers().concat(copies.getInputs())
-                .filter((c) => !INPUT_WHITELIST.some((type) => c instanceof type));
-
-        copies.setInputs(inputs);
-        copies.setOthers(others);
```
Comment (TGCrystal) [inline src/app/digital/models/ioobjects/other/ICData.ts:207]: These functions have a TODO commented above them marking them for deletion once no longer in use. This is the only place in the code where they are used, so the functions themselves could/should also be deleted.

**Review by TGCrystal — CHANGES_REQUESTED** [review]
- [x] Constant Lows/Highs are now given ports on the IC

Aside from that, I didn't find any issues.

File: src/site/pages/digital/src/containers/ICDesigner/index.tsx [hunk 2]
```diff
@@ -39,6 +39,7 @@ import {GetRenderFunc} from "site/digital/utils/Rendering";
 import {CloseICDesigner} from "site/digital/state/ICDesigner";
 
 import "./index.scss";
+import {Event} from "core/utils/Events";
```
[the remark this answers is not in the document]
Comment (LeonMontealegre) [reply src/site/pages/digital/src/containers/ICDesigner/index.tsx:42]: the student has become the master

**Review by TGCrystal — APPROVED** [review]

**Pull request merged**